### ISSUE TITLE
Coral M.2 Accelerator with Dual TPU not showing up on NVIDIA Jetson Nano

### ISSUE THREAD

**Issue by pridkett** [google-coral/edgetpu#339]
There are a few people, myself included, who are experiencing the same problem with [the new M.2 Accelerator with Dual TPUs not showing up on NVIDIA Jetson Nano systems](https://forums.developer.nvidia.com/t/m-2-pcie-google-coral-in-ubuntu/108963/15?u=patrick43). For reference, I was able to get my Single TPU M.2 accelerator to function by [adding in `pcie_aspm=off` to my kernel boot arguments as documented in issue 96](https://github.com/google-coral/edgetpu/issues/96#issuecomment-616566015). So, I know that I have a system with a working M.2 slot and that my libraries are installed correctly.

**lspci output**
```
$  lspci -nn
00:02.0 PCI bridge [0604]: NVIDIA Corporation Device [10de:0faf] (rev a1)
01:00.0 Ethernet controller [0200]: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller [10ec:8168] (rev 15)
```

There is no Coral device that appears. It's my belief that this is at the root of the problem.

**uname -r**
```
$ uname -r
4.9.201-tegra
```

The system is running the standard and most current image for NVIDIA Jetson Nanos.

**lsmod**
```
$  lsmod
Module                  Size  Used by
bnep                   16562  2
fuse                  103969  3
zram                   25920  4
overlay                48718  0
spidev                 13282  0
userspace_alert         5756  0
nvgpu                1581474  18
bluedroid_pm           13912  0
ip_tables              19441  0
x_tables               28951  1 ip_tables
```

gasket and apex were not loaded. Most likely because there's no device showing up on the PCIE bus.

**modinfo gasket**
```
$ modinfo gasket
filename:       /lib/modules/4.9.201-tegra/updates/dkms/gasket.ko
author:         Rob Springer <rspringer@google.com>
license:        GPL v2
version:        1.1.3
description:    Google Gasket driver framework
srcversion:     EE63B2D1E7E5BEBAF51CEE1
depends:
vermagic:       4.9.201-tegra SMP preempt mod_unload modversions aarch64
parm:           dma_bit_mask:int
```

gasket is compiled and works fine with the single TPU accelerator in the system. It's not loaded here because there's no device in `lspci`. Manually loading gasket and apex doesn't help.

**modinfo apex**
```
$  modinfo apex
filename:       /lib/modules/4.9.201-tegra/updates/dkms/apex.ko
author:         John Joseph <jnjoseph@google.com>
license:        GPL v2
version:        1.1
description:    Google Apex driver
srcversion:     58E1862C107AD9271AA7ADA
alias:          pci:v00001AC1d0000089Asv*sd*bc*sc*i*
depends:        gasket
vermagic:       4.9.201-tegra SMP preempt mod_unload modversions aarch64
parm:           allow_power_save:int
parm:           allow_sw_clock_gating:int
parm:           allow_hw_clock_gating:int
parm:           bypass_top_level:int
parm:           trip_point0_temp:int
parm:           trip_point1_temp:int
parm:           trip_point2_temp:int
parm:           hw_temp_warn1:int
parm:           hw_temp_warn2:int
parm:           hw_temp_warn1_en:bool
parm:           hw_temp_warn2_en:bool
parm:           temp_poll_interval:int
```

As with gasket, apex isn't loaded. Which means `/dev/apex_0` does not exist. Manually loading doesn't help here.

I've also attached the output of `lspci -vvv` and `dmesg`, which also make it apparent, the device just doesn't show up on the PCI-E bus  right now.
- [jetson-coral-dmesg.txt](https://github.com/google-coral/edgetpu/files/6157364/jetson-coral-dmesg.txt)
- [jetson-coral-lspci.txt](https://github.com/google-coral/edgetpu/files/6157365/jetson-coral-lspci.txt)

**Comment by Namburger** [comment]
@pridkett hey!
I believe the E-Key on the jetson nano is only for wireless, only the M key is compatible with our module, but it's not dual :/

**Comment by pridkett** [comment]
Thanks @Namburger! If I'm understanding it correctly (and I find the mess of different M.2 form factors confusing) the single TPU which is an A+E key works, probably because of the A key form factor connector or because it needs fewer of the E key lines. The dual TPU version, which is E key only (and has more connectors with traces on them), won't function because of the lack of full fledged M.2 E key support on the Jetson Nano.

If that's the case, then I guess it's off the eBay with the dual TPU M.2 E key. Thanks for the clarification.

I'm going to close this issue. I appreciate the help.

**Comment by DeepDubbed** [comment]
@pridkett 

I've just managed to get my Dual TPU to show-up and work on my Jetson Nano 4GB.

I've basically followed the steps from https://coral.ai/docs/m2/get-started/ 

I've also had to add apply this https://coral.ai/docs/m2/get-started/#troubleshooting-on-linux - basically edit the  /boot/extlinux/extlinux.conf and add pcie_aspm=off but I've aslo added gasket.dma_bit_mask=32 on the same line.

Update: even though I've managed to run inference I don't seem to be able to use the "dual" TPU, meaning the second TPU.

**Comment by pridkett** [comment]
@DeepDubbed that's really interesting - on your system, what is the PCIe device that shows up when you run `lspci -nn`? 

The challenge for me and other people on the NVIDIA forums with similar problems is that the Dual TPU M.2 Coral doesn't even show up when you run `lspci -nn`. The single TPU showed up, but the driver doesn't work until you add `pcie_aspm=off` to `/boot/extlinux/extlinux.conf` (it makes the device show up a little differently and then gasket picks it up).

Also, what does the `gasket.dma_bit_mask=32` setting do? I'm trying to find documentation on that and I can't seem to find anything.

**Comment by DeepDubbed** [comment]
00:01.0 PCI bridge [0604]: NVIDIA Corporation Device [10de:0fae] (rev a1)
00:02.0 PCI bridge [0604]: NVIDIA Corporation Device [10de:0faf] (rev a1)
01:00.0 System peripheral [0880]: Device [1ac1:089a]
02:00.0 Ethernet controller [0200]: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller [10ec:8168] (rev 15)

The gasket.dma_bit_mask doesn't seem to do anything on my nVidia - read about it in the https://coral.ai/docs/m2/get-started/#troubleshooting-on-linux

I had to add pcie_aspm=off because of the errors described in the same section of the doc.

Please note that I don't think it's possible to make both TPUs work due to the limitation of nVidia Jetson Nano - As per their specifications the M.2 E key is assigned to only 1 x4 PCIe. The Dual TPU requires 2 x1 PCIe.

**Comment by pridkett** [comment]
That's the exact same output I get when I have my single TPU installed, which works fine on the Jetson. The dual TPU doesn't even show up in my device tree. We've got the same board revisions too, so that eliminates something.

```
$  lspci -nn
00:01.0 PCI bridge [0604]: NVIDIA Corporation Device [10de:0fae] (rev a1)
00:02.0 PCI bridge [0604]: NVIDIA Corporation Device [10de:0faf] (rev a1)
01:00.0 System peripheral [0880]: Device [1ac1:089a]
02:00.0 Ethernet controller [0200]: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller [10ec:8168] (rev 15)
```

@Namburger: Do you know what the PCI device ID for the dual TPU should be? Is it the same as the single TPU?

**Comment by DeepDubbed** [comment]
I've rechecked my board and it clearly has 2 TPUs on it.
See attachment. ![tpus](https://user-images.githubusercontent.com/79571015/114188543-eb0f9600-9951-11eb-9cb8-8d10b722395a.jpg)

Please note I'm powering my Jetson using a 20W power brick over the barrel and not using USB-C!! It might make a difference in the fact that Jetson has enough power for PCIe ...

**Comment by pridkett** [comment]
Also powering over a 20W power brick on the barrel.

But, I think we have slightly different revisions of the board. Yours says "2027 06-11" and mine says "2027 02-07" at least this is making progress. I don't see any other physical differences in the cards. I'm glad someone has gotten part of the dual TPU working. Maybe just with a different revision.

<img width="469" alt="image" src="https://user-images.githubusercontent.com/160672/114189488-40c74c80-9918-11eb-95cb-565f8cc64dea.png">

**Comment by DeepDubbed** [comment]
I think that's just a serial number, I have another one with 2027 06-10. Both of them from mouser, would expect them to be bulk stocked.

**Comment by mattroos** [comment]
@DeepDubbed and @pridkett I'm experiencing the same problem, except on the AGX Xavier. Like @pridkett, the device is not reported by ```lspci -nn```. Worse yet, I set ```pcie_aspm=off``` in ```/boot/extlinux/extlinux.conf``` and rebooted, only to have the Xavier lock up just after the NVIDIA splash screen appears then disappears. I get a blank screen with a blinking cursor. No combinations of keys, that I've found, will get me to a command line, and I can't see the Xavier on the network or ssh in. Did either of you experience anything similar? Any resolution? Seems I'll have to reformat the SD card and start from scratch. Lots of install and compilation time wasted. :(
